### ISSUE TITLE
SingleChronicleQueueTest.testReadingDocumentWithFirstAMoveWithEpoch started failing consistently on 2018-02-19

### ISSUE THREAD

**Issue by epickrram** [OpenHFT/Chronicle-Queue#436]
Currently ignored.

**Comment by JerryShea** [comment]
Sorry, didn't notice that you were assigned @tomshercliff, and started looking at this as I though it could be relevant to some other problems we were having.
I've created a failing unit test. I can continue with this if you haven't started on it?

**Comment by tomshercliff** [comment]
I have spent some time on this...

The problem is related to the DateTimeFormatter in RollingResourcesCache, which is initialized with a ZoneId based on the specified epochs offset from UTC.

For example, with HOURLY rolling cycles with file format yyyyMMdd-HH:
The resourceFor method formats the Resource filename based on the epochs offset from UTC e.g. if the epoch is specified as [whatever_date]T**02**:08:53.145Z, the filename for cycle 0 (the first cycle) is 19700101-**02**
whereas...
The parseCount0 method calculates the cycle based on the number of hours (i.e. cycles) specified in the filename (i.e. **2** hours for 19700101-**02** giving a cycle of 2) 

These two methods are inconsistent and we see the effects where RollingCycle < Daily e.g. HOURLY or MINUTELY.  (Note: This doesn't seem to be a problem where RollingCycle is >= Daily because the time isn't part of the file format.)

Before fixing, we need to decide which of the methods is correct.  Should the specified epoch dictate the timezone in which the filenames are specified i.e. the time portion of the epoch I specify is my timezones offset from UTC?

To put it another way: if, for an HOURLY rollcycle, I specify my epoch to be "[my_date]T**02**:08:53.145Z" should the filename for the first cycle be "[my_date]-02" or "[my_date]-00"?

**Comment by JerryShea** [comment]
Agree.... I vote for "[my_date]-00"

**Comment by tomshercliff** [comment]
I've refactored RollingResourcesCache and committed with tests:
https://github.com/OpenHFT/Chronicle-Queue/compare/feature/%23436

I'd appreciate a quick review, particularly as I've changed how the epoch being passed in is being used. Previously (for an hourly rollcycle), submitting an epoch of 2018-04-12 02:08:53.145 would result in a filename of 19700101-02 for the first cycle - i.e. the time portion only of the epoch was used to calculate the hour part of the filename.  After my changes, submitting the same epoch will result in a filename of 20180412-00 for the first cycle - i.e. the hour part of the filename taken from offset the epoch specified -which seems more intuitive?

See RollingResourcesCacheTest.shouldCorrectlyConvertCyclesToResourceNamesWithEpoch() for various examples of different epoch values, RollCycles and cycles.

**Comment by JerryShea** [comment]
I would want @RobAustin and @peter-lawrey to review this as I don't know this epoch code at all well.

**Comment by RobAustin** [comment]
@dpisklov could you review this as I'm not going to get time to look at this for about 2 weeks - thanks
